src/include/daos_srv/ras.h

```diff
@@ -0,0 +1,109 @@
+/**
+ * (C) Copyright 2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+/*
+ * RAS event definitions to the used in either data or control planes.
+ */
+
+#ifndef __RAS_H__
+#define __RAS_H__
+
+#define RAS_ID_UNKNOWN_STR "Unknown RAS event"
+#define RAS_SEV_UNKNOWN_STR "Unknown RAS event severity"
+#define RAS_TYPE_UNKNOWN_STR "Unknown RAS event type"
+
+enum ras_event_id {
+	RAS_RANK_EXIT	= 1,
+	RAS_RANK_NO_RESP,
+};
+
+static inline char *
+ras_event_id_enum_to_name(enum ras_event_id id)
+{
+	switch (id) {
+	case RAS_RANK_EXIT:
+		return "daos_rank_exited";
+	case RAS_RANK_NO_RESP:
+		return "daos_rank_no_response";
+	}
+
+	return RAS_ID_UNKNOWN_STR;
```

Comment:
I would rather not maintain a list like that and just put everything in the JSON format itself. IMO, the list of event should only be maintained in the documentation.

Reply:
...pending off-line discussion

Reply:
One potential solution could be to read a json file on control plane start-up parse each entry and register events with pubsub (each entry would map to a concrete events interface implementation and would have a proto message definition to be used on the C side). This seems more complex to me but it might meet your requirements.

The complexity stems from the fact that we need to share the event structures (including boilerplate RAS details and optional. Extended info which is different for specific event type/topic) between C and go code, there are various approaches for this including:
1- protobuf where we generate helper code from proto schema
2- cgo where we would have the definitions in C and access and use the native C types from go 
3- generate native types on both sides from a list of identifiers in a shared location
... (there are probably more)

The current approach (implemented by me with direction from @mjmac) is to use option #3 above with specific event info definitions in proto files, then we have all the auto generated structures for specific event types in C and Go and when raising event we just fill in the details from the shared list of identifiers (in src/daos_srv/ras.h) and populate the auto generated protobuf structures with the relevant event type specific extended info.

Generating event definitions for C and Go from JSON file at runtime or compiletime is not trivial given that we need a schema that understands extended info structure for specific event types. Django framework does this for python and implementation is not trivial.

If we choose to restrict ourselves to a fixed set of event types with predefined event info structures then we could create distinct proto definitions for each event in a JSON blob.